utils/ucl.c

```diff
@@ -0,0 +1,156 @@
+/* Copyright (c) 2015, Cesanta Software
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *       * Redistributions of source code must retain the above copyright
+ *         notice, this list of conditions and the following disclaimer.
+ *       * Redistributions in binary form must reproduce the above copyright
+ *         notice, this list of conditions and the following disclaimer in the
+ *         documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED ''AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL AUTHOR BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stdio.h>
+#include <getopt.h>
+#include <sysexits.h>
+
+#include "ucl.h"
+
+static struct option opts[] = {
+    {"help", no_argument, NULL, 'h'},
+    {"in", required_argument, NULL, 'i' },
+    {"out", required_argument, NULL, 'o' },
+    {"schema", required_argument, NULL, 's'},
+    {"format", required_argument, NULL, 'f'},
+    {0, 0, 0, 0}
+};
+
+void usage(const char *name, FILE *out) {
+  fprintf(out, "Usage: %s [--help] [-i|--in file] [-o|--out file]\n", name);
+  fprintf(out, "    [-s|--schema file] [-f|--format format]\n\n");
+  fprintf(out, "  --help   - print this message and exit\n");
+  fprintf(out, "  --in     - specify input filename "
+          "(default: standard input)\n");
+  fprintf(out, "  --out    - specify output filename "
+          "(default: standard output)\n");
+  fprintf(out, "  --schema - specify schema file for validation\n");
+  fprintf(out, "  --format - output format. Options: ucl (default), "
+          "json, compact_json, yaml\n");
+}
+
+int main(int argc, char **argv) {
+  char ch;
+  FILE *in = stdin, *out = stdout;
+  const char *schema = NULL;
+  unsigned char *buf = NULL;
+  size_t size = 0;
+  struct ucl_parser *parser = NULL;
+  int r;
```

Comment:
`r` is `int` whilst `size` is `size_t`. That seems to be dangerous and irrational, since `fread` returns `size_t`.